test/unit/utils.test.js

```diff
@@ -0,0 +1,78 @@
+'use strict';
+
+const { expect } = require('chai');
+const utils = require('../../src/utils');
+
+describe('utils.js', () => {
+  describe('exports', () => {
+    it('should have toLegacy symbol', () => {
+      expect(utils).to.have.property('toLegacy').that.is.a('symbol');
+    });
+
+    it('should have maybeCallback helper', () => {
+      expect(utils).to.have.property('maybeCallback').that.is.a('function');
+    });
+  });
+
+  describe('maybeCallback', () => {
+    const maybeCallback = utils.maybeCallback;
+    it('should accept to two arguments', () => {
+      expect(maybeCallback).to.have.lengthOf(2);
+    });
+
+    describe('when handling an error case', () => {
+      it('should pass the error to the callback provided', done => {
+        const superPromiseRejection = Promise.reject(new Error('fail'));
+        const result = maybeCallback(superPromiseRejection, (error, result) => {
+          try {
+            expect(result).to.not.exist;
+            expect(error).to.be.instanceOf(Error);
+            return done();
+          } catch (assertionError) {
+            return done(assertionError);
+          }
+        });
+        expect(result).to.be.undefined;
+      });
+
+      it('should return the rejected promise to the caller when no callback is provided', async () => {
+        const superPromiseRejection = Promise.reject(new Error('fail'));
+        const returnedPromise = maybeCallback(superPromiseRejection, null);
+        expect(returnedPromise).to.equal(superPromiseRejection);
+        const thrownError = await returnedPromise.catch(error => error);
+        expect(thrownError).to.be.instanceOf(Error);
+      });
+
+      it('should not modify a rejection error promise', async () => {
+        class MyError extends Error {}
+        const rejection = Promise.reject(new MyError());
+        const thrownError = await maybeCallback(rejection, null).catch(error => error);
+        expect(thrownError).to.be.instanceOf(MyError);
```

Comment:
so this isn't a 100% guarantee, since if we were to extend `MyError` with a new class, it would still be `instanceof MyError`

you could instead instantiate MyError() as readonly, and then pass it to the rejection handler and make sure the error isn't modified;

and we also want to test the same thing for the callback error - that it's not modifying the error in the rejection

Reply:
in order to do that we would need to have access to MyError in order to subclass it, which the only way here from maybeCallback would be to inspect the prototype chain. But I've frozen the error and check instance equality so that should protect us.